lib/msal-common/src/cache/CacheHelpers.ts

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { TemporaryCacheKeys, Constants } from "../utils/Constants";
+import { ICacheStorage } from "./ICacheStorage";
+import { Account } from "../auth/Account";
+import { Authority } from "../auth/authority/Authority";
+import { CodeRequestParameters } from "../server/CodeRequestParameters";
+
+export class CacheHelpers {
+
+    private cacheStorage: ICacheStorage;
+
+    constructor(cacheImpl: ICacheStorage) {
+        this.cacheStorage = cacheImpl;
+    }
+
+    /**
+     * Create acquireTokenAccountKey to cache account object
+     * @param accountId
+     * @param state
+     */
+    generateAcquireTokenAccountKey(accountId: any, state: string): string {
+        return `${TemporaryCacheKeys.ACQUIRE_TOKEN_ACCOUNT}${Constants.RESOURCE_DELIM}${accountId}${Constants.RESOURCE_DELIM}${state}`;
+    }
+
+    /**
+     * Create authorityKey to cache authority
+     * @param state
+     */
+    generateAuthorityKey(state: string): string {
+        return `${TemporaryCacheKeys.AUTHORITY}${Constants.RESOURCE_DELIM}${state}`;
+    }
+
+    /**
+     * @hidden
+     * @ignore
+     *
+     * Sets the cachekeys for and stores the account information in cache
+     * @param account
+     * @param state
+     * @hidden
+     */
+    setAccountCache(account: Account, state: string) {
+        // Cache acquireTokenAccountKey
+        const accountId = account && account.homeAccountIdentifier ? account.homeAccountIdentifier : Constants.NO_ACCOUNT;
+
+        const acquireTokenAccountKey = this.generateAcquireTokenAccountKey(accountId, state);
+        this.cacheStorage.setItem(acquireTokenAccountKey, JSON.stringify(account));
+    }
+
+    /**
+     * @hidden
+     * @ignore
+     *
+     * Sets the cacheKey for and stores the authority information in cache
+     * @param state
+     * @param authority
+     * @hidden
+     */
+    setAuthorityCache(authority: Authority, state: string) {
+        // Cache authorityKey
+        const authorityKey = this.generateAuthorityKey(state);
+        this.cacheStorage.setItem(authorityKey, authority.canonicalAuthority);
+    }
+
+    /**
+     * Updates account, authority, and state in cache
+     * @param serverAuthenticationRequest
+     * @param account
+     * @hidden
+     * @ignore
+     */
+    updateCacheEntries(serverAuthenticationRequest: CodeRequestParameters, account: Account): void {
+        // Cache account and state
+        if (account) {
+            this.setAccountCache(account, serverAuthenticationRequest.state);
+        } else {
+            // Cache the request state
+            this.cacheStorage.setItem(TemporaryCacheKeys.REQUEST_STATE, serverAuthenticationRequest.state);
```

Comment:
Please check the latest dev branch in `msal-core` and update the temp cache with `state` for all entries. You can do this in a separate PR if you want to.